text/0114-dynamic-component-configuration.md

```diff
@@ -0,0 +1,233 @@
+
+```js
+element;
+
+@api
+moduleName;
+
+connectedCallback() {
+	const ctor = await import(this.moduleName);
+	this.element = createComponent('c-lazy-cmp', { is: ctor });
+	// element must be manually inserted into the dom
+	this.template.querySelector('div').appendChild(this.element)
+}
+
+@api
+
+set properties(newProperties) {
+
+this.props = newProperties;
+
+Object.assign(this.element, this.props);
+
+}
+
+get attributes() { return this.props; }
+
+```
+A new tool must be introduced that will allow component authors to configure their components with information retrieved at runtime without relying on maintaining element references or manually manipulating the DOM.
+
+## Proposal
+
+The proposal involves adding a new directive that captures the relevant information to configure a web component at runtime. The assumption is that this directive will be used in cases where the component module is unknown at runtime (therefore used alongside dynamic imports) and zero or more of the following are unknown at runtime: properties, attributes, or event handlers.
+
+The directive `lwc:bind` can retrieve component configuration data via a JavaScript object with a pre-defined shape. Constructor is the only required property in the configuration data.
+
+## General Invariants
+
+* It must not break the reactivity model or the general framework lifecycle
+
+* After the component is loaded, it should function within the lifecycle of the parent component
+
+* Changes to the configuration object must be reflected in the component (ie changing the properties will cause component properties to be set again)
+
+* Changing the constructor on a parent component will also trigger a cascade re-render of any child components 
+
+* Ex: changing the "dashboard" constructor will re-render any child "chart" components
+
+* Will work regardless of whether one uses dynamic imports to retrieve the component constructor
+
+* How a component author retrieves the configurations is unrelated to the proposal
+
+* In cases of invalid component ctor, same behavior as existing `lwc:dynamic` feature
+
+* This should reuse the same mechanism as `lwc:dynamic` and must not require additional development time to add support in environments that currently support `lwc:dynamic`.
+
+## Drawbacks
+
+Identifying the dependencies for static analysis could be difficult especially if the configurations are not derived from @api properties or statically analyzable strings. Additionally, this proposal would enable true dynamic component creation which has been abused in the past - proper guardrails would need to be implemented to prevent improper use of this pattern.
+
+## Adoption Strategy
+
+This feature will be available in Salesforce Lightning Platform and LWR to certain teams depending on their use cases
+
+## Unresolved Questions
```

Comment:
Questions that are not covered by this proposal:

- Can the `lwc:bind` directive be used with other directives, attributes and event handlers?
- If the response to the previous answer is yes, what happens when the `lwc:bind` contains an attribute that is already defined via the template.  For example: `<x-foo lwc:bind={config} class="error"></x-foo>` with `config = { attributes: { class: 'success' } }` ?
- In which order `properties` and `attributes` are set? If the config is defined as `config = { attributes: { class: 'error' }, properties: { className: 'success' }  }` what is the expected output?
- How are attributes values treated? How to add and remove an attribute? What is the semantic of the following attribute values: `null`, `undefined`, `true` and `false`?

Reply:
what about special properties, like `innerHtml`, `outerHtml`, `innerText`, etc? Will they be allowed?

There's already a [inner html binding RFC] (https://github.com/salesforce/lwc-rfcs/pull/15), which states some security concerns. Open question: Does this `lwc:bind` proposal overlaps with `lwc:inner-html`?